packages/alfa-aria/src/name.ts

```diff
@@ -336,48 +337,125 @@ export namespace Name {
   /**
    * @internal
    */
-  export interface State {
+  export class State {
+    private static _empty = new State(Set.empty(), false, false, false);
+
+    public static empty(): State {
+      return this._empty;
+    }
+
+    private readonly _visited: Set<Element>;
+    private readonly _isRecursing: boolean;
+    private readonly _isReferencing: boolean;
+    private readonly _isDescending: boolean;
+
+    private constructor(
+      visited: Set<Element>,
+      isRecursing: boolean,
+      isReferencing: boolean,
+      isDescending: boolean
+    ) {
+      this._visited = visited;
+      this._isRecursing = isRecursing;
+      this._isReferencing = isReferencing;
+      this._isDescending = isDescending;
+    }
+
     /**
      * The elements that have been seen by the name computation so far. This is
      * used for detecting circular references resulting from things such as the
      * `aria-labelledby` attribute and form controls that get their name from
      * a containing `<label>` element.
      */
-    readonly visited: Set<Element>;
+    public get visited(): Set<Element> {
+      return this._visited;
+    }
 
     /**
      * Whether or not the name computation is the result of recursion.
      */
-    readonly isRecursing: boolean;
+    public get isRecursing(): boolean {
+      return this._isRecursing;
+    }
 
     /**
      * Whether or not the name computation is the result of a reference.
      */
-    readonly isReferencing: boolean;
+    public get isReferencing(): boolean {
+      return this._isReferencing;
+    }
 
     /**
      * Whether or not the name computation is descending into a subtree.
      */
-    readonly isDescending: boolean;
-  }
+    public get isDescending(): boolean {
+      return this._isDescending;
+    }
 
-  export namespace State {
-    /**
-     * The initial, empty state of the name computation.
-     */
-    export const empty: State = {
-      visited: Set.empty(),
-      isRecursing: false,
-      isReferencing: false,
-      isDescending: false,
-    };
+    public hasVisited(element: Element): boolean {
+      return this._visited.has(element);
+    }
+
+    public visit(element: Element): State {
+      const visited = this._visited.add(element);
+
+      if (this._visited === visited) {
+        return this;
+      }
```

Comment:
Could get the same logic with probably less computation with
```javascript
if (this.hasVisited(element)) {
  return this;
}
…
```
(would also make the logic more obvious, I feel: if it's already in, do nothing; otherwise, add it.

Reply:
That would require doing _two_ lookups though: First to check if the element is present, and second to add the element if it isn't. I'll add an explanatory comment though.

Reply:
I guess the point I misjudged is the cost of `===` on complex structures (set, map, …)
I naively assumed it needed to compare content (which would also require lookup).

Reply:
Ah, I see! Equality checks for object types are solely based on reference identity, not structure, and so will always be `O(1)`, yes. Structural equality will however be introduced for the proposed record and tuple types, currently at stage 2: https://github.com/tc39/proposal-record-tuple